### PR TITLE
Don't do any DNSSEC record processing when DNSSEC is disabled at build-time.

### PR DESCRIPTION
**Pull request by briansmith**
In the discussoin of #265, @bluejekyll wrote:

> Should we just use the --features=dnssec-* as the method for enabling/disabling dnssec support? 

Implement that.

First, centralize the majority of DNSSEC-specific code into rr::dnssec, so that most of DNSSEC code is in rr::dnssec or into SecureDnsHandle. This makes it easier for somebody reading the code to focus on either the non-DNSSEC logic or the DNSSEC logic, which is helpful for code auditing.

Next, stop building almost all of the DNSSEC-specific logic when the "dnssec" default feature isn't enabled.

### CODE REVIEW

File: proto/src/op/message.rs [hunk 1]
```diff
@@ -526,7 +531,8 @@ impl Message {
                 records.push(record)
             } else {
                 match record.rr_type() {
-                    RecordType::SIG => {
+                    #[cfg(feature = "dnssec")]
+                    RecordType::DNSSEC(DNSSECRecordType::SIG) => {
```
Comment (bluejekyll) [inline proto/src/op/message.rs]: Technically speaking this is SIG0 signing, not DNSSec. i.e. used for dynamic update. Should we create a separate feature for this?

File: proto/src/op/message.rs [hunk 2]
```diff
@@ -606,7 +612,8 @@ impl Message {
         for fin in finals {
             match fin.rr_type() {
                 // SIG0's are special, and come at the very end of the message
-                RecordType::SIG => self.add_sig0(fin),
+                #[cfg(feature = "dnssec")]
+                RecordType::DNSSEC(DNSSECRecordType::SIG) => self.add_sig0(fin),
```
Comment (bluejekyll) [inline proto/src/op/message.rs]: see comment about about SIG0.

File: proto/src/rr/record_data.rs [hunk 3]
```diff
@@ -1189,21 +833,27 @@ mod tests {
...
-            RData::DNSKEY(..) => RecordType::DNSKEY,
             RData::MX(..) => RecordType::MX,
             RData::NS(..) => RecordType::NS,
-            RData::NSEC(..) => RecordType::NSEC,
-            RData::NSEC3(..) => RecordType::NSEC3,
-            RData::NSEC3PARAM(..) => RecordType::NSEC3PARAM,
             RData::NULL(..) => RecordType::NULL,
             RData::OPT(..) => RecordType::OPT,
             RData::PTR(..) => RecordType::PTR,
-            RData::SIG(..) => RecordType::SIG,
```
Comment (bluejekyll) [inline proto/src/rr/record_data.rs]: Again, see comment above about SIG0

File: proto/src/rr/record_type.rs [hunk 4]
```diff
@@ -68,21 +62,11 @@ pub enum RecordType {
...
-    NSEC3PARAM,
     /// RFC 6891	Option
     OPT,
     /// RFC 1035[1]	Pointer record
     PTR,
-    /// RFC 4034	DNSSEC signature: RSASHA256 and RSASHA512, RFC5702
-    RRSIG,
     //  RP,         //	17	RFC 1183	Responsible person
-    /// RFC 2535 (2931)	Signature, to support 2137 Update
-    SIG,
```
Comment (bluejekyll) [inline proto/src/rr/record_type.rs]: See SIG0 comment above.

File: proto/src/rr/rr_set.rs [hunk 5]
```diff
@@ -679,9 +692,11 @@ mod test {
     }
 
     #[test]
+    #[cfg(feature = "dnssec")] // XXX: Should there be a non-DNSSEC version of this?
```
Comment (bluejekyll) [inline proto/src/rr/rr_set.rs]: this only filters based on rfc 6975. so it's valid to disable for dnssec...

**Comment by briansmith** [comment]
> I think the big question I have right now is how to enable SIG0. Based on the current implementation we'd enable DNSSec to get SIG0, which might be reasonable...

IIUC, dnssec-ring or dnssec-openssl is already required to get sig0 working, because those flags control which (if any) crypto implementation to use. So, I don't think this PR changes anything w.r.t. that.

In the future, we may want to enable sig0 signing and/or sig0 verification separate from DNSSEC. I think that could be done pretty straightforwardly on top of this PR.

**Comment by briansmith** [comment]
> Technically speaking this is SIG0 signing, not DNSSec. i.e. used for dynamic update. Should we create a separate feature for this?

I simply took any crypto-looking thing and moved it to the dnssec submodule and the new DNSSEC types. I can see that this PR is technically wrong. My biased opinion is that it isn't so wrong that we need to change it to add new sig0-specific features before it is merged. Perhaps what we should do is document in the `DNSSECRecordType` and `DNSSECRData` types that `SIG` is technically not a DNSSEC thing, but a SIG0 thing, but since they are handled the same way (for now), it's lumped in with them. That way we don't confuse readers of the code.

**Comment by bluejekyll** [comment]
> My biased opinion is that it isn't so wrong that we need to change it to add new sig0-specific features before it is merged.

Yes, by the end of my review I had come to this conclusion.

> Perhaps what we should do is document in the DNSSECRecordType and DNSSECRData types that SIG is technically not a DNSSEC thing, but a SIG0 thing, but since they are handled the same way (for now), it's lumped in with them.

Agreed.

**Pull request merged**